src/protego.py

```diff
@@ -0,0 +1,273 @@
+        if six.PY2:
+            path = quote(unquote(path.encode('utf-8')), safe=safe).decode('utf-8')
+        else:
+            path = quote(unquote(path), safe=safe)
+        path.replace("\n", "%2F")
+        parts = ParseResult('', '', path, parts.params, parts.query, parts.fragment)
+        path = urlunparse(parts)
+        return path
+
+    def allow(self, path):
+        path = self._quote_path(path, safe='/*$')
+        self.rules.append(Rule(field='allow', value=path))
+
+        # if index.html is allowed, we interpret this as / being allowed too.
+        if path.endswith('/index.html'):
+            self.allow(path[:-10] + '$')
+
+    def disallow(self, path):
+        path = self._quote_path(path, safe='/*$')
+        self.rules.append(Rule(field='disallow', value=path))
+
+    def _parser_match(self, pattern, url):
+        pattern = re.sub(r'\*+', '*', pattern)
+        s = re.split(r'([$*])', pattern)
+        for index, substr in enumerate(s):
+            if substr not in ['*', '$']:
+                s[index] = re.escape(substr)
+            else:
+                s[index] = s[index].replace('*', '.*')
+        pattern = '^' + ''.join(s)
+        if re.findall(pattern, url):
+            return True
+        return False
+
+    def allowed(self, url):
+        if not self.rules_sorted:
+            self.rules = sorted(self.rules, key=lambda r: (len(r.value), r.field == 'allow'), reverse=True)
+            self.rules_sorted = True
+        url = self._quote_path(url)
+        allowed = True
+        for rule in self.rules:
+            if self._parser_match(rule.value, url):
+                if rule.field == 'disallow':
+                    allowed = False
+                break
+        return allowed
+
+    def set_crawl_delay(self, delay):
+        try:
+            delay = float(delay)
+        except ValueError:
+            delay = None
+        self.crawl_delay = delay
+
+    def get_crawl_delay(self):
+        return self.crawl_delay
+
+    def set_request_rate(self, rate, time_period):
+        requests, seconds = rate.split('/')
+        requests = int(requests)
+        time_unit = seconds[-1].lower()
+        seconds = int(seconds[:-1])
+        if time_unit == 'm':
+            seconds *= 60
+        elif time_unit == 'h':
+            seconds *= 3600
+        elif time_unit == 'd':
+            seconds *= 86400
+
+        start_time = None
+        end_time = None
+        if time_period:
+            start_time, end_time = time_period.split('-')
+            start_time = time(int(start_time[:2]), int(start_time[-2:]))
+            end_time = time(int(end_time[:2]), int(end_time[-2:]))
+
+        self.req_rate = RequestRate(requests, seconds, start_time, end_time)
+
+    def get_request_rate(self):
+        return self.req_rate
+
+
+class Protego:
+
+    def __init__(self):
+        self.user_agents = {}
+        self.host = None
+        self.sitemap_list = []
+        self.total_line_seen = 0
```

Comment:
If these variables are only for tests, it might be a good idea to mark them as private.

I see you’ve already marked some members as private, but you might want to review Protego’s API in general, and mark classes, methods and attributes that are not meant for end users to use as private.

That way, you make a statement that those API members are considered implementation details, and you might change or remove them in the future without aiming to keep backward compatibility. Users who use them cannot blame you if their code stops working after such changes.